remote_shell_service.go

```diff
@@ -0,0 +1,222 @@
+package main
+
+import (
+	"bufio"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/subtle"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/gliderlabs/ssh"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	"golang.org/x/term"
+)
+
+type Console struct {
+	stdInLock  sync.Mutex
+	stdInPipe  io.Writer
+	stdOutPipe io.Reader
+	stdErrPipe io.Reader
+
+	sessionLock    sync.Mutex
+	remoteSessions map[uuid.UUID]ssh.Session
+}
+
+func makeConsole(stdin io.Writer, stdout io.Reader, stderr io.Reader) Console {
+	return Console{
+		stdInPipe:      stdin,
+		stdOutPipe:     stdout,
+		stdErrPipe:     stderr,
+		remoteSessions: map[uuid.UUID]ssh.Session{},
+	}
+}
+
+func (c *Console) WriteToStdIn(p []byte) (n int, err error) {
+	c.stdInLock.Lock()
+	n, err = c.stdInPipe.Write(p)
+	c.stdInLock.Unlock()
```

Comment:
I wonder if these are already thread-safe, but then the lock certainly doesn't hurt in this usage pattern.

Reply:
What little research I've been able to do says that there's no requirement for any given io.Writer to be thread-safe. But even if it is, how does it lock? Is it possible to get interleaved writes, which breaks the intended behavior in this case? 

This is a case where I'd rather be overly safe than have to chase down some race condition or the like 6 months from now.

Reply:
Agreed and I'm also so far not finding docs that explicitly say either way. If it was (like Java does in some cases with I/O streams), the unit of string/bytes passed would be atomically written.